### PR TITLE
Refactoring and UI updates

### PR DESCRIPTION
**Pull request by john-royal**
This includes some feature additions, UI improvements, bug fixes, and some fairly significant refactoring. Some changes might not be agreeable to everyone, so I'm happy to modify this PR as needed.

### Views
* Added ability to view all posts by a certain author from `PostsList`/`PostRow`
* Redesigned `NewPostForm` and relocated from `MainTabView` to a sheet presented from the `PostsList`
* Fixed post titles not being shown by `PostRow`
* Added error, empty, and loading state UI to `PostsList`
* Refined appearance of post and profile images in `ProfileView` and `PostRow`
* Made `MainTabView` more focused by relocating auth-related views to `AppView`, which is now the root view of the app
* Significantly reorganized and streamlined `PostsList` and `CommentsList`

### View Models
* Replaced async dispatch in `TaskViewModel` with `MainActor` attribute
* Renamed `PostData` to `PostViewModel` for consistency with other view models

### Models
* Updated `User.testUser` and `Post.testPost` to improve SwiftUI previews
* Added partial objects to `Comment` and `Post` to help simplify `CommentService` and `PostService` APIs

### Services
* Renamed `UserService` to `AuthService` for consistency with `AuthViewModel`
* Significantly refactored `AuthService`, `CommentService`, and `PostService` and refined their public APIs
* Removed no longer needed version numbering from collection references (e.g. used the `posts` collection instead of `posts_v2`)

### Utilities
* Consolidated Firestore-related utilities and extensions in new `Firestore+Extensions` file
* Relocated image upload logic from `PostService` and `AuthService` to new `Storage+Extensions` file

### CODE REVIEW

**Review by benstone1 — CHANGES_REQUESTED** [review]
Left a bunch of comments, but most are focused on style and me getting a better understanding of how things work.  Great work on this!  Love the structure that you have with it, learners are going to get a ton out of these tutorials.

Additionally, we got the updated design specs in.  You can find them at the link [here](https://drive.google.com/drive/folders/1zRJcSw3ynsHI7JjTgXmAM_2iQe-nPavb).  You should also go ahead and tweak the views so that they match closer to the designs.  Because we're not focusing as much on the UI for this skill path, you should feel free to do whatever makes the code simplest while using the design for inspiration.

File: FirebaseTestProject/Models/User.swift [hunk 1]
```diff
@@ -7,23 +7,35 @@
...
-    
-    static let testUser = User(id: "0000000000000000000000000000", name: "Jane Doe")
+}
+
+// MARK: - Test
+
+extension User {
+    static let testUser = User(
+        id: "0000000000000000000000000000",
+        name: "Jamie Harris",
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Models/User.swift:38]: Glad you asked! I like using race and gender neutral names where possible, to be more inclusive. Jamie Harris is my go-to, but feel free to LMK if you have any other ideas!

File: FirebaseTestProject/Models/User.swift [hunk 2]
```diff
@@ -7,23 +7,35 @@
...
         }
     }
+    
     private var imageURLString = ""
     
-    init(id: String, name: String) {
+    init(id: String, name: String, imageURL: URL? = nil) {
         self.id = id
         self.name = name
+        self.imageURL = imageURL
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Models/User.swift:29]: This sets the `imageURL` to `nil`. Please see my comment above.

File: FirebaseTestProject/Services/AuthService.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+// MARK: - AuthService
+
+struct AuthService: AuthServiceProtocol {
+    var auth = Auth.auth()
+    var imagesReference = Storage.storage().reference().child("images/users")
+    
+    func createAccount(name: String, email: String, password: String) async throws -> User {
+        let result = try await auth.createUser(withEmail: email, password: password)
+        
+        let profileUpdate = result.user.createProfileChangeRequest()
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/AuthService.swift:31]: Good idea! I'll make that change.

File: FirebaseTestProject/Services/CommentService.swift [hunk 4]
```diff
@@ -19,51 +39,28 @@ struct CommentService {
         self.post = post
         self.user = postService.user
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/CommentService.swift:40]: Nice catch! I’ll rework this.

File: FirebaseTestProject/Services/AuthService.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        profileUpdate.photoURL = imageURL
+        try await profileUpdate.commitChanges()
+        
+        return User(from: user)
+    }
+}
+
+private extension User {
+    init(from user: FirebaseAuth.User) {
+        id = user.uid
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/AuthService.swift:72]: I thought it was cleaner this way, but now that you mention it, I think the clarity of explicitly assigning to `self` is preferable. I'll add that in.

File: FirebaseTestProject/Services/PostService.swift [hunk 6]
```diff
@@ -8,69 +8,125 @@
...
+
+// MARK: - PostFilter
+
+enum PostFilter {
+    case author(User)
+    case favorites
+}
+
+extension PostServiceProtocol {
+    func fetchPosts(matching filter: PostFilter?) async throws -> [Post] {
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/PostService.swift:44]: Why? Seems a little redundant given that `fetchPosts(matching:)` with no filter parameter becomes simply `fetchPosts()`, which already exists. Am I missing something?

File: FirebaseTestProject/Services/PostService.swift [hunk 7]
```diff
@@ -8,69 +8,125 @@
...
     var favoritesReference = Firestore.firestore().collection("favorites")
     var imagesReference = Storage.storage().reference().child("images/posts")
     
-    private var postsQuery: Query {
-        postsReference.order(by: "timestamp", descending: false)
-    }
-    private var favoritesQuery: Query {
-        favoritesReference.whereField("userID", isEqualTo: user.id)
+    func fetchPosts() async throws -> [Post] {
+        async let postsTask = postsReference
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/PostService.swift:65]: Fair point! I’ll try to figure out a better solution and get back to you.

File: FirebaseTestProject/Views/CommentRow.swift [hunk 8]
```diff
@@ -7,72 +7,71 @@
...
-private extension Comment {
-    var displayTimestamp: String {
-        let formatter = DateFormatter()
-        formatter.dateStyle = .short
-        formatter.timeStyle = .short
-        return formatter.string(from: timestamp)
+// MARK: - DeleteViewModifier
+
+private extension CommentRow {
+    struct DeleteViewModifier: ViewModifier {
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Views/CommentRow.swift:41]: Good idea! I’ll add the extension.

**Review by john-royal — COMMENTED** [review]
Thank you for the feedback! I left some replies to answer your questions, but I’ll get to work ASAP on the requested changes.

File: FirebaseTestProject/Services/AuthService.swift [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    func updateProfileImage(_ image: UIImage) async throws -> User {
+        guard let user = auth.currentUser else {
+            preconditionFailure("Cannot update image because there is no signed in user")
+        }
+        
+        let imageReference = imagesReference.child("\(user.uid).jpg")
+        let imageURL = try await imageReference.uploadImage(image)
+        
+        let profileUpdate = user.createProfileChangeRequest()
+        profileUpdate.photoURL = imageURL
```
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Services/AuthService.swift:63]: Haha, never mind then!  I assumed it was something that we owned.

File: FirebaseTestProject/Services/PostService.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Services/PostService.swift:44]: I guess the question is why accept a `nil` `PostFilter` at all when a `fetchPosts()` exists.

File: FirebaseTestProject/ViewModels/PostViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+//
+
+import Foundation
+
+@MainActor class PostViewModel: ObservableObject {
+    @Published var posts: [Post] = []
+    @Published var state: LoadingState = .loading
+    
+    private let postService: PostServiceProtocol
+    private let filter: PostFilter?
```
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/ViewModels/PostViewModel.swift:15]: I'm good with it this way :+1:

File: FirebaseTestProject/Views/PostsList.swift [hunk 11]
```diff
@@ -7,65 +7,143 @@
...
+                    .navigationTitle("Favorites")
+            }
+        }
+    }
+}
+
+// MARK: - RouterView
+
+private extension PostsList {
+    @MainActor struct RouterView: View {
```
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Views/PostsList.swift:102]: Interesting, let me play around with it locally.

File: FirebaseTestProject/Models/User.swift [hunk 12]
```diff
@@ -7,23 +7,35 @@
...
     var name: String
-    var imageURL: URL {
+    var imageURL: URL? {
         get {
-            URL(string: imageURLString) ?? Bundle.main.url(forResource: "ProfileImagePlaceholder", withExtension: "png")!
+            URL(string: imageURLString)
         }
         set {
-            imageURLString = newValue.absoluteString
+            imageURLString = newValue?.absoluteString ?? ""
```
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Models/User.swift:20]: Ah, that makes sense to me.  As long as we remember to explain this context to users, this should be great then.

File: FirebaseTestProject/Models/User.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Models/User.swift:38]: Jamie works for me!

File: FirebaseTestProject/Services/AuthService.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Services/AuthService.swift:72]: It seems a lot handwavier in Swift 5.5, so feel free to go either way on it.

File: FirebaseTestProject/ViewModels/CommentViewModel.swift [hunk 13]
```diff
@@ -9,36 +9,33 @@ import Foundation
...
-        case loading
-        case error
-        case loaded
-    }
-    
-    private let commentService: CommentService
+    private let commentService: CommentServiceProtocol
     
-    init(commentService: CommentService) {
+    init(commentService: CommentServiceProtocol) {
```
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/ViewModels/CommentViewModel.swift:16]: That's fair, we can leave it as is then.

File: FirebaseTestProject/Services/AuthService.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/AuthService.swift]: I agree with the spirit of this suggestion, but I tried changing it to `profileUpdateRequest` and found the result to be a bit verbose:
```swift
let profileUpdateRequest = result.user.createProfileChangeRequest()
profileUpdateRequest.displayName = name
try await profileUpdateRequest.commitChanges()
```
That’s why I instead renamed the variable `changeRequest`. This was [used in Firebase’s official sample projects](https://github.com/firebase/quickstart-ios/blob/c59bf0e8eb70582085282c90e835fa770146e948/authentication/AuthenticationExample/ViewControllers/UserViewController.swift#L145), and I thought it struck a good balance between clarity and concision. What do you think?

File: FirebaseTestProject/Services/AuthService.swift [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    
+    func updateProfileImage(_ image: UIImage) async throws -> User {
+        guard let user = auth.currentUser else {
+            preconditionFailure("Cannot update image because there is no signed in user")
+        }
+        
+        let imageReference = imagesReference.child("\(user.uid).jpg")
+        let imageURL = try await imageReference.uploadImage(image)
+        
+        let profileUpdate = user.createProfileChangeRequest()
```
[the remark this answers is not in the document]
Comment (john-royal) [reply FirebaseTestProject/Services/AuthService.swift]: I renamed it to `changeRequest`. Please see the note above.

**Review by john-royal — COMMENTED** [review]
I made some updates and left some replies. I’ll also update the UI, but I haven’t finished working on that yet.

I tried to address every comment you left, but some big changes include:
* Replacing the `FirestoreConvertable` protocol with Firebase’s native `Codable` support. This means [our workaround for saving user and post image URLs is no longer necessary](https://github.com/benstone1/SocialCademy/commit/8b717d8015d61a6006801ff6c9b581327a326b26#diff-2ddd49e57145415229cf3539cc302152138cec3dce7335acb160b23ed3fa6114) 🥳.
* Renaming the `Partial` object for posts and comments to `EditableFields`. Method parameter names have also been updated to `editablePost` and `editableComment`.

A few threads are still open. I left replies, but feel free to give me your thoughts whenever you’re ready!

File: FirebaseTestProject/Models/Comment.swift [hunk 15]
```diff
@@ -7,18 +7,36 @@
...
+        content: "Great job!",
+        author: .testUser,
+        id: "00000000000000000000"
+    )
+}
+
+// MARK: - Partial
+
+extension Comment {
+    struct Partial {
```
[the remark this answers is not in the document]
Comment (benstone1) [reply FirebaseTestProject/Models/Comment.swift]: Yeah, I think that name works :+1:

**Pull request merged**